stdlib/public/core/Diffing.swift

```diff
@@ -192,615 +165,168 @@ extension BidirectionalCollection where Element : Equatable {
-///     case let .removed(s):
-///       // use `s`
-///     case let .inserted(t):
-///       // use `t`
-///     case .matched: continue
-///     }
-///   }
-///
-/// [lcs]: http://en.wikipedia.org/wiki/Longest_common_subsequence_problem
-/// [ses]: http://en.wikipedia.org/wiki/Edit_distance
-///
-/// - Note: `CollectionChanges` holds a reference to state used to run the
-///     difference algorithm, which can be exponentially larger than the
-///     changes themselves.
-fileprivate struct _CollectionChanges<SourceIndex, TargetIndex>
-where SourceIndex: Comparable, TargetIndex: Comparable {
-  fileprivate typealias Endpoint = (x: SourceIndex, y: TargetIndex)
-
-  /// An encoding of change elements as an array of index pairs stored in
-  /// `pathStorage[pathStartIndex...]`.
-  ///
-  /// This encoding allows the same storage to be used to run the difference
-  /// algorithm, report the result, and repeat in place using
-  /// `formChanges`.
-  ///
-  /// The collection of changes between XABCD and XYCD is:
-  ///
-  ///   [.match(0..<1, 0..<1), .remove(1..<3), .insert(1..<2),
-  ///    .match(3..<5, 2..<4)]
-  ///
-  /// Which gets encoded as:
-  ///
-  ///   [(0, 0), (1, 1), (3, 1), (3, 2), (5, 4)]
-  ///
-  /// You can visualize it as a two-dimensional path composed of remove
-  /// (horizontal), insert (vertical), and match (diagonal) segments:
-  ///
-  ///     X A B C D
-  ///   X \ _ _
-  ///   Y    |
-  ///   C     \
-  ///   D     \
-  ///
-  private var pathStorage: [Endpoint]
-
-  /// The index in `pathStorage` of the first segment in the difference path.
-  private var pathStartIndex: Int
-
-  /// Creates a collection of changes from a difference path.
-  fileprivate init(
-    pathStorage: [Endpoint], pathStartIndex: Int
-  ) {
-    self.pathStorage = pathStorage
-    self.pathStartIndex = pathStartIndex
-  }
-
-  /// Creates an empty collection of changes, i.e. the changes between two
-  /// empty collections.
-  private init() {
-    self.pathStorage = []
-    self.pathStartIndex = 0
-  }
-}
-
-extension _CollectionChanges {
-  /// A range of elements removed from the source, inserted in the target, or
-  /// that the source and target have in common.
-  fileprivate enum Element {
-    case removed(Range<SourceIndex>)
-    case inserted(Range<TargetIndex>)
-    case matched(Range<SourceIndex>, Range<TargetIndex>)
-  }
-}
+// MARK: Internal implementation
 
-extension _CollectionChanges: RandomAccessCollection {
-  fileprivate typealias Index = Int
+// V is a rudimentary type made to represent the rows of the triangular matrix type used by the Myer's algorithm
+//
+// This type is basically an array that only supports indexes in the set `stride(from: -d, through: d, by: 2)` where `d` is the depth of this row in the matrix
+// `d` is always known at allocation-time, and is used to preallocate the structure.
+fileprivate struct V {
 
-  fileprivate var startIndex: Index {
-    return 0
-  }
+  private var a: [Int]
+#if DEBUG
```

Comment:
Thanks for pointing that out! If it becomes a problem we can remove it, but it's a very useful guardrail since mixing odd and even indexes into a single instance of `_V` is extremely bad.